features/dnsresolver/netty/src/main/java/org/opennms/netmgt/dnsresolver/netty/CaffeineDnsCache.java

```diff
@@ -0,0 +1,370 @@
+        if (maxTtl == 0) {
+            return e;
+        }
+        resolveCache.cache(appendDot(hostname), e, Math.max(minTtl, (int) Math.min(maxTtl, ptrRecord.timeToLive())));
+        return e;
+    }
+
+    /**
+     * Negative caching for failures (or empty reverse lookups).
+     */
+    @Override
+    public DnsCacheEntry cache(String hostname, DnsRecord[] additionals, Throwable cause, EventLoop loop) {
+        checkNotNull(hostname, "hostname");
+        checkNotNull(cause, "cause");
+        checkNotNull(loop, "loop");
+
+        DefaultExtendedDnsCacheEntry e = new DefaultExtendedDnsCacheEntry(hostname, cause);
+        if (negativeTtl == 0 || !emptyAdditionals(additionals)) {
+            return e;
+        }
+
+        resolveCache.cache(appendDot(hostname), e, negativeTtl);
+        return e;
+    }
+
+    @Override
+    public String toString() {
+        return MoreObjects.toStringHelper(getClass())
+                .add("minTtl", minTtl)
+                .add("maxTtl", maxTtl)
+                .add("negativeTtl", negativeTtl)
+                .add("cached resolved hostname=", resolveCache.size())
+                .toString();
+    }
+
+    public long getSize() {
+        return resolveCache.size();
+    }
+
+    private enum CacheMetric {
+        SIZE("cacheSize", c -> (Gauge<Long>) c.resolveCache::size),
+        MAX_SIZE("cacheMaxSize", c -> (Gauge<Long>) c::maxSize),
+        EVICTION_COUNT("cacheEvictionCount", c -> (Gauge<Long>) c.resolveCache.stats()::evictionCount),
+        HITS("cacheHits", c -> c.cacheHits),
+        MISSES("cacheMisses", c -> c.cacheMisses);
+
+        String name;
+        Function<CaffeineDnsCache, Metric> metricSupplier;
+
+        CacheMetric(String name, Function<CaffeineDnsCache, Metric> metricSupplier) {
+            this.name = Objects.requireNonNull(name);
+            this.metricSupplier = Objects.requireNonNull(metricSupplier);
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public Metric getMetric(CaffeineDnsCache cache) {
+            return metricSupplier.apply(cache);
+        }
+    }
+
+    public void registerMetrics(MetricRegistry metrics) {
+        for (CacheMetric cm : CacheMetric.values()) {
+            metrics.register(cm.getName(), cm.getMetric(this));
+        }
+    }
+
+    public void unregisterMetrics(MetricRegistry metrics) {
+        for (CacheMetric cm : CacheMetric.values()) {
+            metrics.remove(cm.getName());
+        }
+    }
+
+    private static final class DefaultExtendedDnsCacheEntry implements ExtendedDnsCacheEntry {
+        private final String hostname;
+        private final String hostnameFromPtrRecord;
+        private final InetAddress address;
+        private final Throwable cause;
+
+        DefaultExtendedDnsCacheEntry(String hostname, InetAddress address) {
+            this.hostname = hostname;
+            this.address = address;
+            cause = null;
```

Comment:
Nit: `this` consistency.

Reply:
Done. I tried not using `this.` unless necessary, but it does "look" better when using `this.` consistently 😄